tests/drivers/uart/uart_async_api/src/test_uart_async.c

```diff
@@ -127,7 +127,7 @@ ZTEST_USER(uart_async_single_read, test_single_read)
 	zassert_not_equal(memcmp(tx_buf, rx_buf, 5), 0,
 			  "Initial buffer check failed");
 
-	uart_rx_enable(uart_dev, rx_buf, 10, 50 * USEC_PER_MSEC);
+	uart_rx_enable(uart_dev, rx_buf, 5, 50 * USEC_PER_MSEC);
```

Comment:
RX buffer here is intentionally larger than TX buffer because particular UART drivers may or may not release the RX buffer after the first transfer (this is the whole point of #57847 where commit eb44414af99a44447afa52bb81ac0767f2ff791e was introduced) and this test is supposed to check if both transfers are successful, regardless of whether the initial RX buffer is released after the first transfer.
For the same reason, the `memcmp` at the end needs to check `tdata.last_rx_buf`, not `tdata.rx_buf`, as some drivers may receive both transfers to the first RX buffer (that `rx_buf` on the stack) and some may use the second one (`tdata.rx_buf`).

Reply:
> For the same reason, the `memcmp` at the end needs to check `tdata.last_rx_buf`, not `tdata.rx_buf`, as some drivers may receive both transfers to the first RX buffer (that `rx_buf` on the stack) and some may use the second one (`tdata.rx_buf`).

If the buffer is not released, then tdata.last_rx_buf does not have to point to &rx_buf[5]. It can point to somewhere in the range &rx_buf[5]  to &rx_buf[9]. It depends on the timeout in uart_rx_enable() call.

Reply:
An alternative could be to add a sleep after the first uart_tx call.

Reply:
> If the buffer is not released, then tdata.last_rx_buf does not have to point to &rx_buf[5]. It can point to somewhere in the range &rx_buf[5] to &rx_buf[9]. It depends on the timeout in uart_rx_enable() call.

Right. Currently, the test assumes that there won't be gaps longer than 50 ms between consecutive transmitted bytes, but it actually cannot be guaranteed. So indeed, this `last_rx_buf` pointer is not the best idea for this second check and this could be improved.

> An alternative could be to add a sleep after the first uart_tx call.

I'm not sure I understand what you mean. The test already waits 100 ms before the second TX: https://github.com/zephyrproject-rtos/zephyr/blob/312b1f9888b5e0e27e3cc511131b60bb0f171948/tests/drivers/uart/uart_async_api/src/test_uart_async.c#L137-L138

Reply:
Yes, you are right.

Reply:
> Right. Currently, the test assumes that there won't be gaps longer than 50 ms between consecutive transmitted bytes, but it actually cannot be guaranteed, so indeed, this `last_rx_buf` pointer is not the best idea for that second transfer check and this could be improved.

And if the timeout occurs before all the 5 bytes are received in the second transfer, we'll generate two RX_RDY calls, which will make validating the test messier.